File: android/src/main/java/expo/modules/googlemobileads/RewardedAd.kt
package expo.modules.googlemobileads

import android.app.Activity
import com.google.android.libraries.ads.mobile.sdk.common.AdLoadCallback
import com.google.android.libraries.ads.mobile.sdk.common.AdRequest
import com.google.android.libraries.ads.mobile.sdk.common.AdValue
import com.google.android.libraries.ads.mobile.sdk.common.FullScreenContentError
import com.google.android.libraries.ads.mobile.sdk.common.LoadAdError
import com.google.android.libraries.ads.mobile.sdk.rewarded.OnUserEarnedRewardListener
import com.google.android.libraries.ads.mobile.sdk.rewarded.RewardItem
import com.google.android.libraries.ads.mobile.sdk.rewarded.RewardedAdEventCallback
import expo.modules.kotlin.AppContext
import java.util.concurrent.atomic.AtomicReference
import com.google.android.libraries.ads.mobile.sdk.rewarded.RewardedAd as GmaRewardedAd

/**
 * The rewarded format. Registered with JS as `RewardedAd`. The GMA type of the same name is
 * imported as `GmaRewardedAd`, as in [InterstitialAd].
 */
class RewardedAd(
  appContext: AppContext,
  adUnitId: String,
  requestOptions: Map<String, Any?>?
) : FullScreenAd(appContext, adUnitId, requestOptions) {

  @Volatile
  private var ad: GmaRewardedAd? = null

  /**
   * What this ad offers, snapshotted from `getRewardItem()` when it loads. Confirmed `@NotNull` in
   * the AAR and readable before the ad is shown, exactly like iOS's `adReward` — so, exactly like
   * on iOS, **its presence says nothing about whether the reward was earned.** It is here so a
   * prompt can say "watch an ad for 10 coins" before showing anything.
   */
  @Volatile
  var reward: Map<String, Any?>? = null
    private set

  /**
   * The reward the user actually earned, or null if [OnUserEarnedRewardListener] never fired.
   *
   * This is the Android half of the parity that matters most. Android hands the reward straight to
   * the listener as a `RewardItem` parameter, so "was a reward earned?" is simply "is this
   * non-null?". iOS reaches the identical observable result the hard way: its
   * `GADUserDidEarnRewardHandler` takes **no arguments**, so it has to latch a boolean and read the
   * pre-populated `adReward` at that moment. Both platforms therefore resolve `show()` with the
   * reward only when the handler fired, and with `null` for a user who dismissed without earning
   * one — which is a real-money distinction no compiler or unit test would catch.
   *
   * An `AtomicReference` for the same reason as the base's `showPromise`: written from an SDK
   * callback with no documented thread, read by [showResult] from the dismissal callback and by
   * nothing else.
   */
  private val earnedReward = AtomicReference<Map<String, Any?>?>(null)

  override fun loadAd(request: AdRequest, loadId: Long) {
    GmaRewardedAd.load(
      request,
      object : AdLoadCallback<GmaRewardedAd> {
        // `loadId` is captured here for the same reason as in InterstitialAd.kt: the SDK gives a
        // result no identity of its own.
        override fun onAdLoaded(ad: GmaRewardedAd) {
          postToMain { onAdReady(ad, loadId) }
        }

        override fun onAdFailedToLoad(adError: LoadAdError) {
          postToMain { handleLoadFailed(adError, loadId) }
        }
      }
    )
  }

  /** Main thread. */
  private fun onAdReady(loaded: GmaRewardedAd, loadId: Long) {
    if (isStaleLoadResult(loadId) || shouldDiscardLoadResult) {
      // Superseded by a later `load()`, or the object can no longer accept any result: released, a
      // show is in flight, or this ad has already been shown. See the same guard in
      // InterstitialAd.kt for why installing here would hang the show promise — and note that the
      // ad destroyed is the one that just arrived, never the one that is presenting.
      //
      // This is also what closes the reward hole: it is the only path that could have installed a
      // new ad while an earlier one's reward listener could still fire, or carried an
      // `earnedReward` from presentation #1 into presentation #2. An ad can only have been
      // presented if `showPromise != null` (during) or `status == "shown"` (after), and both are
      // discarded here.
      loaded.destroy()
      return
    }
    // Nothing should be here: `load()` tears the previous ad down before it bumps the load id, and
    // only the current id reaches this line. Kept unconditional because what it clears matters —
    // any previous ad (which would leak without `destroy()`) and, critically, `earnedReward`, so no
    // reward recorded for an earlier ad can be read out by this one. Safe only because the guard
    // above has ruled out a presentation in flight.
    tearDownAd()
    loaded.adEventCallback = object : RewardedAdEventCallback {
      override fun onAdShowedFullScreenContent() = handleShowed()

      override fun onAdDismissedFullScreenContent() = handleDismissed()

      override fun onAdFailedToShowFullScreenContent(fullScreenContentError: FullScreenContentError) =
        handleFailedToShow(fullScreenContentError)

      override fun onAdImpression() = handleImpression()

      override fun onAdClicked() = handleClicked()

      override fun onAdPaid(value: AdValue) = handlePaid(value)

      // onAdMetadataChanged is left at its default no-op: ad metadata is not part of the JS API.
    }
    ad = loaded
    // Snapshot the offered reward now, so the JS-side `reward` getter never has to touch the ad.
    // Called as functions, not properties: the AAR's Kotlin metadata declares `getRewardItem()`
    // and `getResponseInfo()` as functions, so no `loaded.rewardItem` is synthesized.
    reward = loaded.getRewardItem().toMap()
    handleLoaded(loaded.getResponseInfo())
  }

  override fun presentAd(activity: Activity): Boolean {
    val current = ad ?: return false
    // Spelled out as an object expression rather than a lambda: `OnUserEarnedRewardListener` is a
    // Kotlin interface in the AAR, and SAM conversion from a Kotlin lambda only applies to Java
    // interfaces and to Kotlin `fun interface`s.
    current.show(
      activity,
      object : OnUserEarnedRewardListener {
        override fun onUserEarnedReward(rewardItem: RewardItem) {
          // Unlike iOS, the reward arrives as a parameter. Recorded before it is emitted, so a
          // listener that also reads `show()`'s eventual result cannot see the two disagree.
          val earned = rewardItem.toMap()
          earnedReward.set(earned)
          emit("earnedReward", earned)
        }
      }
    )
    return true
  }

  /** Resolves `showAsync()` with the reward **only if the listener actually fired**. */
  override fun showResult(): Map<String, Any?>? = earnedReward.get()

  override fun tearDownAd() {
    ad?.let {
      it.adEventCallback = null
      it.destroy()
    }
    ad = null
    // Reached at the start of load() and before a load result replaces an existing ad, so the
    // per-presentation record is cleared here rather than only at release. A stale `earnedReward`
    // surviving into a second ad resolves show() with a reward nobody earned on this presentation.
    //
    // This is not belt-and-braces, and an earlier comment here claiming the `"shown"` guard in
    // load() made it unreachable was wrong: before fix round 1 the load callback installed a new ad
    // without ever routing through here, so a result landing after a presentation reached exactly
    // that bug without going through load() at all.
    earnedReward.set(null)
    reward = null
  }
}

/**
 * `RewardItem.getAmount()` is an `Int` here and an `NSDecimalNumber` on iOS; the JS-side
 * `AdReward.amount` is a plain number, which both fit losslessly for real reward quantities.
 */
private fun RewardItem.toMap(): Map<String, Any?> = mapOf(
  "type" to type,
  "amount" to amount
)

File: ios/RewardedAd.swift
import ExpoModulesCore
import GoogleMobileAds
import UIKit

/// The rewarded format. Registered with JS as `RewardedAd`.
///
/// Named `FullScreenRewardedAd` for the same reason as `FullScreenInterstitialAd`: GMA's
/// `GADRewardedAd` is `NS_SWIFT_NAME(RewardedAd)`, so reusing that name here would shadow it.
final class FullScreenRewardedAd: FullScreenAd {
  /// The loaded ad. Main thread only.
  private var ad: RewardedAd?

  // MARK: - The earned-reward latch
  //
  // `GADUserDidEarnRewardHandler` is `typedef void (^)(void)` — it takes **no parameters**. The
  // reward is read off `GADRewardedAd.adReward`, which is `nonnull` and already populated
  // *before* the ad is ever presented (that is what makes the JS-side `reward` property useful
  // for a "watch an ad for 10 coins" prompt).
  //
  // So a non-nil `adReward` is NOT evidence that anything was earned. The only signal is "the
  // zero-argument handler fired". `_didEarnReward` is that signal, and `showResult()` returns the
  // reward only when it is set — otherwise a user who dismissed the ad after two seconds would be
  // granted the reward anyway. Neither the compiler nor a unit test catches that; it is a
  // real-money bug.
  //
  // Guarded by its own lock rather than by "it is all on main": `reward` is read from the JS
  // thread through a `Property` getter. `_reward` is snapshotted at load time so neither reader
  // has to touch `ad`.
  private let rewardLock = NSLock()
  private var _didEarnReward = false
  private var _reward: [String: Any]?

  /// What this ad offers, readable before it is shown. Its presence says nothing about whether
  /// the reward was earned — only `showAsync()`'s resolved value does.
  var reward: [String: Any]? {
    rewardLock.lock()
    defer { rewardLock.unlock() }
    return _reward
  }

  override func loadAd(request: Request) {
    RewardedAd.load(with: adUnitId, request: request) { [weak self] ad, error in
      // Hopped to main for the same reason as the interstitial: the load completion handler is
      // not documented as main-thread, and everything below is main-thread state.
      DispatchQueue.main.async {
        self?.handleLoadCompletion(ad: ad, error: error)
      }
    }
  }

  private func handleLoadCompletion(ad: RewardedAd?, error: Error?) {
    if let error {
      handleLoadFailed(error)
      return
    }
    guard let ad else {
      handleLoadFailed(missingAdError())
      return
    }
    if shouldDiscardLoadResult {
      // Released, a show is in flight, or this ad has already been shown. See the same guard in
      // InterstitialAd.swift for why installing here would hang the show promise.
      //
      // It also closes the last latch hole: this is the only path that could have installed a new
      // `_reward` snapshot while an earlier ad's reward handler was still able to fire, which
      // would have returned the new ad's reward for the old ad's earn. An ad can only have been
      // presented if `showPromise != nil` (during) or `status == "shown"` (after), and both are
      // discarded here — so a snapshot can never be installed after any presentation.
      //
      // Dropping this ad is complete cleanup: its delegate and paid closure are still nil, so
      // returning releases the last reference and ARC frees it.
      return
    }
    // Two loads can overlap (the second `load()` clears `ad`, but the first request may still land
    // afterwards), so an earlier ad can be sitting here. This clears its delegate and paid closure
    // and resets the earned-reward latch. Safe here only because the guard above has ruled out a
    // presentation in flight.
    tearDownAd()
    ad.fullScreenContentDelegate = delegateProxy
    ad.paidEventHandler = { [weak self] value in
      self?.emitPaid(value)
    }
    self.ad = ad
    // Snapshot the offered reward now. It is available pre-show, and taking a copy keeps the
    // JS-side `reward` getter off both `ad` and the main thread.
    let reward = rewardToDictionary(ad.adReward)
    rewardLock.lock()
    _reward = reward
    rewardLock.unlock()
    handleLoaded(responseInfo: ad.responseInfo)
  }

  @MainActor
  override func presentAd(from viewController: UIViewController) -> Bool {
    guard let ad else { return false }
    ad.present(from: viewController) { [weak self] in
      // Zero arguments — see the latch comment above.
      self?.handleUserDidEarnReward()
    }
    return true
  }

  private func handleUserDidEarnReward() {
    rewardLock.lock()
    _didEarnReward = true
    let reward = _reward
    rewardLock.unlock()
    // `_reward` is set whenever the ad loaded, and an ad can only be presented once it has, so
    // this is always non-nil here. Guarded rather than force-unwrapped anyway.
    if let reward {
      emit(event: "earnedReward", payload: reward)
    }
  }

  /// Resolves `showAsync()` with the reward **only if the handler actually fired**.
  override func showResult() -> [String: Any]? {
    rewardLock.lock()
    defer { rewardLock.unlock() }
    return _didEarnReward ? _reward : nil
  }

  /// Also **resets the latch**. `_didEarnReward` belongs to one presentation of one ad object; a
  /// flag left set from a previous ad would make the next `show()` resolve with a reward nobody
  /// earned. `load()` calls this before starting a new request, and `handleLoadCompletion` calls
  /// it before replacing an ad, so no load can ever begin with the latch already set.
  override func tearDownAd() {
    ad?.fullScreenContentDelegate = nil
    ad?.paidEventHandler = nil
    ad = nil
    rewardLock.lock()
    _didEarnReward = false
    _reward = nil
    rewardLock.unlock()
  }
}

/// `GADAdReward.amount` is an `NSDecimalNumber`; the JS-side `AdReward.amount` is a plain number
/// (Android's `RewardItem.getAmount()` is an `Int`). Rewards are whole-number quantities in
/// practice, so the double conversion is lossless for real values.
private func rewardToDictionary(_ reward: AdReward) -> [String: Any] {
  [
    "type": reward.type,
    "amount": reward.amount.doubleValue,
  ]
}
